### ISSUE TITLE
Using Hyper and Relm? (http examples broken)

### ISSUE THREAD

**Issue by iBelieve** [antoyo/relm#63]
Is it possible to use Hyper with Relm? The example in [relm-easyfibers](https://github.com/antoyo/relm/blob/master/relm-easyfibers/examples/http/src/main.rs) appears to be outdated as the `Relm.handle()` method seems to be gone, apparently from moving from tokio to futures-glib. Is the a current working solution to Http futures from Relm? I need to be able to call some Http endpoints and update the Relm UI with the results.

**Comment by superlou** [comment]
Also interested, though I'm hoping to respond to pushed ws-rs updates. :+1:

**Comment by antoyo** [comment]
This is not possible since the update to `0.10`.
I was tired to wait for the updates of the futures/tokio crates and now I still have to wait for [this issue](https://github.com/hyperium/hyper/issues/1274), so I made a release which does not work with most networking code.
Also, the hyper crate depends on `tokio-proto` which is going to be deprecated.
I'm not sure about the future directions relm will take, but it might even remove the tokio support and provide a `relm-futures` crate or something.

**Comment by antoyo** [comment]
You currently have two options:

one would be to use `hyper` from another thread and send messages to the UI thread.

By most networking code, I meant that we cannot use code that needs a [Handle](https://docs.rs/tokio-core/0.1.9/tokio_core/reactor/struct.Handle.html).
You would need to use code that works with [futures-glib](https://github.com/antoyo/futures-glib-rs) (and there's actually [an example of an HTTP request](https://github.com/antoyo/futures-glib-rs/blob/master/examples/tcp.rs)).
A bigger example of using `futures-glib` is in my [fg-uds crate](https://github.com/antoyo/fg-uds-rs).

As I told you, I might stop using futures-glib crate in relm in the future (because I found out that tokio is over-engineered ([see here for some reasons](https://www.reddit.com/r/rust/comments/6s4axu/asynchronous_rust_complaints_suggestions/)) and hard to debug) so you might need to update your code later if you choose this approach.

**Comment by lnicola** [comment]
Hmm, that doesn't look so bad actually. Just to make sure, that works without polling on the main thread, right?

**Comment by antoyo** [comment]
@lnicola Not sure exactly what you mean.
The glib event loop works with the `poll` syscall (at least on Unix).
You'll send a message from the other thread to the main thread and relm adds a source to the glib event loop will poll this new source as well.
That won't block the main thread, though; that's just how event loops work (like the one of tokio).

**Comment by lnicola** [comment]
Sorry for not being clear. I think most multi-threaded `gtk-rs` examples I've seen polled a channel in a `glib::timeout_add` callback, which is not ideal. The better way is to use a single-shot `glib::idle_add`, as I'm sure you know.

If you're using the latter, I guess removing `futures`/`tokio` integration isn't that much of an issue. Perhaps it would have been worth mentioning in your blog post that `relm::Channel` exists. I only looked at some `relm` blog posts and docs, but I missed it.

**Comment by antoyo** [comment]
You're right, polling the channel in `timeout_add()` is not a good idea (relm used to do that for tokio in the very first versions -_- ).
Today, it uses a `GSource`, so it's similar in some ways to `idle_add`.

I'll write a tutorial which will explain, among other things, how to send a message from another thread (i.e. using a `Channel`).
